linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4746,12 +4850,52 @@ private[optimizer] abstract class OptimizerCore(
 
   private def foldAsInstanceOf(arg: PreTransform, tpe: Type)(
       implicit pos: Position): PreTransform = {
-    if (isSubtype(arg.tpe.base, tpe))
+    def mayRequireUnboxing: Boolean =
+      arg.tpe.isNullable && !isNullableType(tpe)
+
+    if (semantics.asInstanceOfs == CheckedBehavior.Unchecked && !mayRequireUnboxing)
+      foldCast(arg, tpe)
+    else if (isSubtype(arg.tpe.base, tpe))
       arg
     else
       AsInstanceOf(finishTransformExpr(arg), tpe).toPreTransform
   }
 
+  private def foldCast(arg: PreTransform, tpe: Type)(
+      implicit pos: Position): PreTransform = {
+    def default(arg: PreTransform, newTpe: RefinedType): PreTransform =
+      PreTransTree(Transient(Cast(finishTransformExpr(arg), tpe)), newTpe)
+
+    def castLocalDef(arg: PreTransform, newTpe: RefinedType): PreTransform = arg match {
+      case PreTransMaybeBlock(bindingsAndStats, PreTransLocalDef(localDef)) =>
+        val refinedLocalDef = localDef.withRefinedType(newTpe)
+        if (refinedLocalDef ne localDef)
```

Comment:
This line really threw me off, but it feels it is because of the naming of `withRefinedType`: It doesn't always refine the type. Maybe call it `withRefinedTypeMaybe` or `tryWithRefinedType`?

Reply:
Good point. Renamed to `tryWithRefinedType`.